File: my-folder/2236-maximum-twin-sum-of-a-linked-list/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:
        # reverse the first half of the linkedlist
        slow, fast = head, head
        reverse = None
        while fast and fast.next:
            fast = fast.next.next
            reverse, reverse.next, slow = slow, reverse, slow.next
        
        first = reverse

        second = slow
        max_val = -math.inf
        while first:
            if first.val + second.val > max_val:
                max_val = first.val + second.val
            first = first.next
            second = second.next
        return max_val

File: my-folder/2236-maximum-twin-sum-of-a-linked-list/solution.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {number}
 */
var pairSum = function(head) {
    let slow = head;
    let fast = head;
    let reverse = null;

    while(fast && fast.next){
        fast = fast.next.next;
        let temp = slow.next;
        slow.next = reverse;
        reverse = slow;
        slow = temp;
    }

    let first = reverse;
    let second = slow;

    let max_val = Number.NEGATIVE_INFINITY;

    while(first) {
        if(max_val < first.val + second.val) {
            max_val = first.val + second.val;
        }
        first = first.next;
        second = second.next;
    }

    return max_val;
    
};
